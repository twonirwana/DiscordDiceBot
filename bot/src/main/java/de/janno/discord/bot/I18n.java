package de.janno.discord.bot;

import com.google.common.collect.ImmutableList;
import de.janno.discord.connector.api.slash.CommandLocaleChoice;
import de.janno.discord.connector.api.slash.CommandLocaleDescription;
import de.janno.discord.connector.api.slash.CommandLocaleName;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.text.StringEscapeUtils;

import java.text.MessageFormat;
import java.util.*;

@Slf4j
public final class I18n {
    private final static String MESSAGES_KEY = "botMessages";

    private I18n() {
    }

    public static String getMessage(String key, Locale locale) {
        return StringEscapeUtils.unescapeJava(ResourceBundle.getBundle(MESSAGES_KEY, locale).getString(key));
    }

    public static String getMessage(String key, Locale locale, Object... arguments) {
        return MessageFormat.format(getMessage(key, locale), arguments);
    }

    public static List<Locale> getAdditionalLanguage() {
        return List.of(Locale.GERMAN, Locale.of("pt","BR"));
    }

    public static List<Locale> allSupportedLanguage() {
        return ImmutableList.<Locale>builder()
                .add(Locale.ENGLISH)
                .addAll(getAdditionalLanguage())
                .build();
    }

    private static List<LocaleString> allNoneEnglishMessagesValues(String key) {
        return getAdditionalLanguage().stream()
                .map(l -> new LocaleString(l, getMessage(key, l)))
                .filter(m -> !Objects.equals(m.value(), getMessage(key, Locale.ENGLISH))) //remove all locals that are equal to the default english one
                .toList();
    }

    public static List<CommandLocaleName> allNoneEnglishMessagesNames(String key) {
        return allNoneEnglishMessagesValues(key).stream()
                .map(lv -> new CommandLocaleName(lv.locale(), lv.value()))
                .toList();
    }

    public static List<CommandLocaleDescription> allNoneEnglishMessagesDescriptions(String key) {
        return allNoneEnglishMessagesValues(key).stream()
                .map(lv -> new CommandLocaleDescription(lv.locale(), lv.value()))
                .toList();
    }

    public static List<CommandLocaleChoice> allNoneEnglishMessagesChoices(String key) {
        return allNoneEnglishMessagesValues(key).stream()
                .map(lv -> new CommandLocaleChoice(lv.locale(), lv.value()))
                .toList();
    }

    public static List<CommandLocaleDescription> allNoneEnglishDescriptionsWithKeys(String key, String... keys) {
        return getAdditionalLanguage().stream()
                .map(l -> new CommandLocaleDescription(l, getMessage(key, l, Arrays.stream(keys)
                        .map(k -> getMessage(k, l))
                        .toArray(Object[]::new))))
                .filter(m -> !Objects.equals(m.getDescription(), getMessage(key, Locale.ENGLISH))) //remove all locals that are equal to the default english one
                .toList();
    }

    private record LocaleString(Locale locale, String value){}

}
