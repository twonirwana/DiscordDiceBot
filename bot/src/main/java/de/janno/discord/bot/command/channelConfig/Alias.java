package de.janno.discord.bot.command.channelConfig;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.NonNull;
import lombok.Value;
import org.jetbrains.annotations.NotNull;

@Value
public class Alias {

    @NonNull String name;
    @NonNull String value;

    @JsonCreator
    public Alias(@NotNull @JsonProperty("name") String name, @NotNull @JsonProperty("value") String value) {
        this.name = name;
        this.value = value;
    }

    @Override
    public String toString() {
        return "%s->%s".formatted(name, value);
    }
}
