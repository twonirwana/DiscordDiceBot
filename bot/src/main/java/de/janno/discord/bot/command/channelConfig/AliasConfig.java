package de.janno.discord.bot.command.channelConfig;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.NonNull;
import lombok.Value;
import org.jetbrains.annotations.NotNull;

import java.util.List;

@Value
public class AliasConfig {

    @NonNull List<Alias> aliasList;

    @JsonCreator
    public AliasConfig(@NotNull @JsonProperty("aliasList") List<Alias> aliasList) {
        this.aliasList = aliasList;
    }

}
