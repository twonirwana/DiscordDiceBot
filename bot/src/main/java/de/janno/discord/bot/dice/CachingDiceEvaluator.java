package de.janno.discord.bot.dice;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.cache.CacheBuilder;
import com.google.common.cache.CacheLoader;
import com.google.common.cache.LoadingCache;
import de.janno.evaluator.dice.DiceEvaluator;
import de.janno.evaluator.dice.ExpressionException;
import de.janno.evaluator.dice.Roller;
import de.janno.evaluator.dice.random.NumberSupplier;
import io.micrometer.core.instrument.Gauge;
import io.micrometer.core.instrument.Tags;
import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;

import java.util.concurrent.ExecutionException;

import static io.micrometer.core.instrument.Metrics.globalRegistry;

@Slf4j
public class CachingDiceEvaluator {

    private final LoadingCache<String, RollerOrError> diceRollerCache;
    private final DiceEvaluator diceEvaluator;

    public CachingDiceEvaluator(NumberSupplier numberSupplier, int maxNumberOfDice, int cacheSize) {
        diceEvaluator = new DiceEvaluator(numberSupplier, maxNumberOfDice);
        diceRollerCache = CacheBuilder.newBuilder()
                .maximumSize(cacheSize)
                .recordStats()
                .build(new CacheLoader<>() {
                    @Override
                    public @NonNull RollerOrError load(@NonNull String expression) {
                        try {
                            log.debug("create roller for: {}", expression);
                            Roller roller = diceEvaluator.buildRollSupplier(expression);
                            roller.roll();
                            return new RollerOrError(expression, roller, true, null);
                        } catch (ExpressionException | ArithmeticException e) {
                            return new RollerOrError(expression, null, false, e.getMessage());
                        }
                    }
                });
        Gauge.builder("diceEvaluator.cache", diceRollerCache::size).tags(Tags.of("stats", "size")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().requestCount()).tags(Tags.of("stats", "requests")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().hitCount()).tags(Tags.of("stats", "hit")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().missCount()).tags(Tags.of("stats", "miss")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().totalLoadTime()).tags(Tags.of("stats", "totalLoadTime")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().averageLoadPenalty()).tags(Tags.of("stats", "averageLoadTime")).register(globalRegistry);
    }

    public @NonNull RollerOrError get(@NonNull String expression) {
        try {
            return diceRollerCache.get(expression);
        } catch (ExecutionException e) {
            throw new IllegalStateException(e);
        }
    }

    @VisibleForTesting
    public long getCacheSize() {
        return diceRollerCache.size();
    }
}
