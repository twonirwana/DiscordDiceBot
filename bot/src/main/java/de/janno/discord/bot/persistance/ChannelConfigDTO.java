package de.janno.discord.bot.persistance;

import lombok.AllArgsConstructor;
import lombok.NonNull;
import lombok.Value;

import java.util.UUID;

@Value
@AllArgsConstructor
public class ChannelConfigDTO {

    @NonNull
    UUID configUUID;
    Long guildId;
    long channelId;
    Long userId;
    @NonNull
    String commandId;
    @NonNull
    String configClassId;
    @NonNull
    String config;
}
