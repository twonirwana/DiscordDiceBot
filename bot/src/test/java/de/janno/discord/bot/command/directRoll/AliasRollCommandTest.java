package de.janno.discord.bot.command.directRoll;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import de.janno.discord.bot.I18n;
import de.janno.discord.bot.command.channelConfig.Alias;
import de.janno.discord.bot.command.channelConfig.AliasConfig;
import de.janno.discord.bot.dice.CachingDiceEvaluator;
import de.janno.discord.bot.persistance.ChannelConfigDTO;
import de.janno.discord.bot.persistance.Mapper;
import de.janno.discord.bot.persistance.PersistenceManager;
import de.janno.discord.bot.persistance.PersistenceManagerImpl;
import de.janno.discord.connector.api.AutoCompleteAnswer;
import de.janno.discord.connector.api.AutoCompleteRequest;
import org.apache.commons.io.FileUtils;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.io.File;
import java.io.IOException;
import java.util.List;
import java.util.Locale;
import java.util.UUID;
import java.util.stream.Stream;

import static de.janno.discord.bot.command.channelConfig.AliasHelper.CHANNEL_ALIAS_CONFIG_TYPE_ID;
import static de.janno.discord.bot.command.channelConfig.AliasHelper.USER_ALIAS_CONFIG_TYPE_ID;
import static org.assertj.core.api.Assertions.assertThat;

@ExtendWith(SnapshotExtension.class)

public class AliasRollCommandTest {
    private PersistenceManager persistenceManager;
    private Expect expect;
    private AliasRollCommand underTest;

    static Stream<Arguments> generateAllLocaleData() {
        return I18n.allSupportedLanguage().stream()
                .map(Arguments::of);
    }

    @BeforeEach
    void setup() throws IOException {
        File cacheDirectory = new File("imageCache/");
        if (cacheDirectory.exists()) {
            FileUtils.cleanDirectory(cacheDirectory);
        }
        persistenceManager = new PersistenceManagerImpl("jdbc:h2:mem:" + UUID.randomUUID(), null, null);
        underTest = new AliasRollCommand(persistenceManager, new CachingDiceEvaluator((minExcl, maxIncl) -> 1, 1000, 0));
    }

    @ParameterizedTest(name = "{index} locale={0}")
    @MethodSource("generateAllLocaleData")
    void testHelp(Locale userLocale) {
        expect.scenario(userLocale.toString()).toMatchSnapshot(underTest.getHelpMessage(userLocale));
    }

    @Test
    public void getCommandDefinition() {
        expect.toMatchSnapshot(underTest.getCommandDefinition());
    }

    @Test
    public void getId() {
        assertThat(underTest.getCommandId()).isEqualTo("a");
    }

    @Test
    void autoCompleteNoAlias() {

        List<AutoCompleteAnswer> res = underTest.getAutoCompleteAnswer(new AutoCompleteRequest("alias_or_expression", null, List.of()), Locale.ENGLISH, 1L, 1L);

        assertThat(res.stream().map(AutoCompleteAnswer::getName)).containsExactly("No alias configured in this channel, add them with `/channel_config alias save`");
        assertThat(res.stream().map(AutoCompleteAnswer::getValue)).containsExactly("help");
    }

    @Test
    void autoCompleteValid() {
        createAlias();

        List<AutoCompleteAnswer> res = underTest.getAutoCompleteAnswer(new AutoCompleteRequest("alias_or_expression", null, List.of()), Locale.ENGLISH, 1L, 1L);

        assertThat(res.stream().map(AutoCompleteAnswer::getName)).containsExactly("a", "b", "c");
        assertThat(res.stream().map(AutoCompleteAnswer::getValue)).containsExactly("a", "b", "c");
    }

    @Test
    void autoCompleteValid_noUserAlias() {
        createAlias();

        List<AutoCompleteAnswer> res = underTest.getAutoCompleteAnswer(new AutoCompleteRequest("alias_or_expression", null, List.of()), Locale.ENGLISH, 1L, 3L);

        assertThat(res.stream().map(AutoCompleteAnswer::getName)).containsExactly("b", "c");
        assertThat(res.stream().map(AutoCompleteAnswer::getValue)).containsExactly("b", "c");
    }

    @Test
    void autoCompleteFilter() {
        createAlias();

        List<AutoCompleteAnswer> res = underTest.getAutoCompleteAnswer(new AutoCompleteRequest("alias_or_expression", "b", List.of()), Locale.ENGLISH, 1L, 1L);

        assertThat(res.stream().map(AutoCompleteAnswer::getName)).containsExactly("b");
        assertThat(res.stream().map(AutoCompleteAnswer::getValue)).containsExactly("b");
    }

    @Test
    void autoCompleteFilterNotFound() {
        createAlias();

        List<AutoCompleteAnswer> res = underTest.getAutoCompleteAnswer(new AutoCompleteRequest("alias_or_expression", "bc", List.of()), Locale.ENGLISH, 1L, 1L);

        assertThat(res).isEmpty();
    }

    private void createAlias() {
        persistenceManager.saveChannelConfig(new ChannelConfigDTO(UUID.randomUUID(),
                1L,
                1L,
                1L,
                "channel_config",
                USER_ALIAS_CONFIG_TYPE_ID,
                Mapper.serializedObject(new AliasConfig(List.of(new Alias("a", "1d6"), new Alias("b", "2d6"))))));

        persistenceManager.saveChannelConfig(new ChannelConfigDTO(UUID.randomUUID(),
                1L,
                1L,
                null,
                "channel_config",
                CHANNEL_ALIAS_CONFIG_TYPE_ID,
                Mapper.serializedObject(new AliasConfig(List.of(new Alias("b", "1d8"), new Alias("c", "2d8"))))));

        persistenceManager.saveChannelConfig(new ChannelConfigDTO(UUID.randomUUID(),
                1L,
                2L,
                1L,
                "channel_config",
                USER_ALIAS_CONFIG_TYPE_ID,
                Mapper.serializedObject(new AliasConfig(List.of(new Alias("z", "1d10"), new Alias("a", "1d10"), new Alias("b", "1d10"))))));

        persistenceManager.saveChannelConfig(new ChannelConfigDTO(UUID.randomUUID(),
                1L,
                2L,
                null,
                "channel_config",
                CHANNEL_ALIAS_CONFIG_TYPE_ID,
                Mapper.serializedObject(new AliasConfig(List.of(new Alias("z", "1d10"), new Alias("b", "1d10"), new Alias("c", "1d10"))))));

        persistenceManager.saveChannelConfig(new ChannelConfigDTO(UUID.randomUUID(),
                1L,
                2L,
                2L,
                "channel_config",
                USER_ALIAS_CONFIG_TYPE_ID,
                Mapper.serializedObject(new AliasConfig(List.of(new Alias("y", "1d12"), new Alias("a", "1d12"), new Alias("b", "1d12"))))));
    }

}
