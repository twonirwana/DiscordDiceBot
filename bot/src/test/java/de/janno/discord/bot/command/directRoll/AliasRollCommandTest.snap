de.janno.discord.bot.command.directRoll.AliasRollCommandTest.getCommandDefinition=[
  {
    "name": "a",
    "description": "Roll of an alias or a dice expression. Add alias with `/channel_config alias save`",
    "nameLocales": [ ],
    "descriptionLocales": [
      {
        "locale": "de",
        "description": "Direkter Wurf eines Alias oder eines Ausdrucks. Neue Aliase mit `/channel_config alias save`"
      },
      {
        "locale": "pt_BR",
        "description": "fórmula um alias ou de uma expressão de dados. Adicionar alias com `/channel_config alias save`"
      },
      {
        "locale": "fr",
        "description": "Lancer d'un alias ou d'une expression de dé. Ajouter un alias avec `/channel_config alias save`"
      }
    ],
    "options": [
      {
        "type": "STRING",
        "name": "alias_or_expression",
        "nameLocales": [ ],
        "description": "Roll of an alias or a dice expression. Add alias with `/channel_config alias save`",
        "descriptionLocales": [
          {
            "locale": "de",
            "description": "Direkter Wurf eines Alias oder eines Ausdrucks. Neue Aliase mit `/channel_config alias save`"
          },
          {
            "locale": "pt_BR",
            "description": "fórmula um alias ou de uma expressão de dados. Adicionar alias com `/channel_config alias save`"
          },
          {
            "locale": "fr",
            "description": "Lancer d'un alias ou d'une expression de dé. Ajouter un alias avec `/channel_config alias save`"
          }
        ],
        "required": true,
        "choices": [ ],
        "options": [ ],
        "autoComplete": true
      }
    ],
    "integrationTypes": [
      "GUILD_INSTALL",
      "USER_INSTALL"
    ]
  }
]


de.janno.discord.bot.command.directRoll.AliasRollCommandTest.testHelp[de]=[
  {
    "descriptionOrContent": "Schreib `/a`und ein Alias oder Würfelausdruck. Verfügbare Alias werden im Autocomplete gezeigt, es ist aber auch möglich normale Ausrücke zu würfeln. Aliase können mit `/channel_config alias` angelegt oder gelöscht werden. Antwort Format und Würfelbilder können mit `/channel_config save_direct_roll_config` konfiguriert werden.\n```\n| Name                 | Notation   | Example            |\r\n|----------------------|------------|--------------------|\r\n| Regular Dice         | d          | d6, 3d6            |\r\n| Custom Dice          | d[x,y,...] | 3d[Head,Tail]      |\r\n| Exploding Dice       | d!         | 4d!6               |\r\n| Exploding Add Dice   | d!!        | 4d!!6              |\r\n| Keep Highest         | k          | 3d6k2              |\r\n| Keep Lowest          | l          | 3d6l2              |\r\n| Sum                  | =          | 3d6=               |\r\n| Add to List          | +          | 3d6+2d12           |\r\n| Concatenate          | _          | 3d6 _ 'dmg'        |\r\n| Negative add to List | -          | 3d6-2d12 or -2d6   |\r\n| Decimal Divide       | //         | 2/6                |\r\n| Divide               | /          | 12/6               |\r\n| Multiply             | *          | 12*d6              |\r\n| Modulo               | mod        | d6 mod 2           |\r\n| Equal Filter         | ==         | 3d6==3             |\r\n| Greater Then Filter  | >          | 3d6>3              |\r\n| Lesser Then Filter   | <          | 3d6<3              |\r\n| Greater Equal Filter | >=         | 3d6>=3             |\r\n| Lesser Equal Filter  | <=         | 3d6<=3             |\r\n| Lesser Equal Filter  | <=         | 3d6<=3             |\r\n| Count                | c          | 3d6<3c             |\r\n| Multiple Rolls       | ,          | 1d6,2d10           |\r\n| Repeat               | x          | 3x2d6              |\r\n| Repeat List          | r          | 3r2d6              |\r\n| Reroll               | rr         | 3d10 rr 1          |\r\n| Equal                | =?         | d6=?6              |\r\n| Greater              | >?         | d6>?5              |\r\n| Greater Equal        | >=?        | d6>=?5             |\r\n| Lower                | <?         | d6<?4              |\r\n| Lower Equal          | <=?        | d6<=?4             |\r\n| In                   | in         | d6 in [1/3/5]      |\r\n| And                  | &&         | d6>?1 && d6<5      |\r\n| Or                   | ||         | d6>?1 || d6<5      |\r\n| Negate               | !          | !d6>?1             |\r\n| Color                | col        | d6 col 'red'       |\r\n| Tag                  | tag        | d6 tag 'marked'    |\r\n| Brackets             | ()         | (2d4=)d6           |\r\n| Text                 | ''         | 'Damage ' + d6     |\r\n| Sort asc             | asc()      | asc(10d10)         |\r\n| Sort desc            | desc()     | desc(10d10)        |\r\n| Min                  | min()      | min(3d4, 8)        |\r\n| Max                  | max()      | max(3d4, 8)        |\r\n| Chancel              | chancel()  | chancel(8d10,10,1) |\r\n| Replace              | replace()  | replace(6d10,1,2)  |\r\n| Color On             | colorOn()  | colorOn(d6,6,'red')|\r\n| Explode              | exp()      | exp(d6,1,2)        |\r\n| If                   | if()       | if(d6=?6,'Y','N')  |\r\n| Group Count          | groupC()   | groupC(20d6)       |\r\n| Concatenate          | concat()   | concat('Att:',d20) |\r\n| Value                | val()      | val('$1',10d10)    |\r\n\n```",
    "fields": [
      {
        "name": "Beispiel",
        "value": "`/a alias_or_expression: attack` for an alias with the name `attack`",
        "inline": false
      },
      {
        "name": "Beispiel",
        "value": "`/a alias_or_expression: 1d6`",
        "inline": false
      },
      {
        "name": "Vollständige Dokumentation",
        "value": "https://github.com/twonirwana/DiscordDiceBot",
        "inline": false
      },
      {
        "name": "Discord Server für Hilfe und Neuigkeiten",
        "value": "https://discord.gg/e43BsqKpFr",
        "inline": false
      }
    ],
    "componentRowDefinitions": [ ],
    "type": "EMBED",
    "userReference": false
  }
]


de.janno.discord.bot.command.directRoll.AliasRollCommandTest.testHelp[en]=[
  {
    "descriptionOrContent": "Type `/a` and an alias or dice expression. Available alias will be presented as autocomplete, but it is possible to roll all expression. `/channel_config alias` can be used to add and delete alias. The output type, dice images etc. can be configuration with `/channel_config save_direct_roll_config`\n```\n| Name                 | Notation   | Example            |\r\n|----------------------|------------|--------------------|\r\n| Regular Dice         | d          | d6, 3d6            |\r\n| Custom Dice          | d[x,y,...] | 3d[Head,Tail]      |\r\n| Exploding Dice       | d!         | 4d!6               |\r\n| Exploding Add Dice   | d!!        | 4d!!6              |\r\n| Keep Highest         | k          | 3d6k2              |\r\n| Keep Lowest          | l          | 3d6l2              |\r\n| Sum                  | =          | 3d6=               |\r\n| Add to List          | +          | 3d6+2d12           |\r\n| Concatenate          | _          | 3d6 _ 'dmg'        |\r\n| Negative add to List | -          | 3d6-2d12 or -2d6   |\r\n| Decimal Divide       | //         | 2/6                |\r\n| Divide               | /          | 12/6               |\r\n| Multiply             | *          | 12*d6              |\r\n| Modulo               | mod        | d6 mod 2           |\r\n| Equal Filter         | ==         | 3d6==3             |\r\n| Greater Then Filter  | >          | 3d6>3              |\r\n| Lesser Then Filter   | <          | 3d6<3              |\r\n| Greater Equal Filter | >=         | 3d6>=3             |\r\n| Lesser Equal Filter  | <=         | 3d6<=3             |\r\n| Lesser Equal Filter  | <=         | 3d6<=3             |\r\n| Count                | c          | 3d6<3c             |\r\n| Multiple Rolls       | ,          | 1d6,2d10           |\r\n| Repeat               | x          | 3x2d6              |\r\n| Repeat List          | r          | 3r2d6              |\r\n| Reroll               | rr         | 3d10 rr 1          |\r\n| Equal                | =?         | d6=?6              |\r\n| Greater              | >?         | d6>?5              |\r\n| Greater Equal        | >=?        | d6>=?5             |\r\n| Lower                | <?         | d6<?4              |\r\n| Lower Equal          | <=?        | d6<=?4             |\r\n| In                   | in         | d6 in [1/3/5]      |\r\n| And                  | &&         | d6>?1 && d6<5      |\r\n| Or                   | ||         | d6>?1 || d6<5      |\r\n| Negate               | !          | !d6>?1             |\r\n| Color                | col        | d6 col 'red'       |\r\n| Tag                  | tag        | d6 tag 'marked'    |\r\n| Brackets             | ()         | (2d4=)d6           |\r\n| Text                 | ''         | 'Damage ' + d6     |\r\n| Sort asc             | asc()      | asc(10d10)         |\r\n| Sort desc            | desc()     | desc(10d10)        |\r\n| Min                  | min()      | min(3d4, 8)        |\r\n| Max                  | max()      | max(3d4, 8)        |\r\n| Chancel              | chancel()  | chancel(8d10,10,1) |\r\n| Replace              | replace()  | replace(6d10,1,2)  |\r\n| Color On             | colorOn()  | colorOn(d6,6,'red')|\r\n| Explode              | exp()      | exp(d6,1,2)        |\r\n| If                   | if()       | if(d6=?6,'Y','N')  |\r\n| Group Count          | groupC()   | groupC(20d6)       |\r\n| Concatenate          | concat()   | concat('Att:',d20) |\r\n| Value                | val()      | val('$1',10d10)    |\r\n\n```",
    "fields": [
      {
        "name": "Example",
        "value": "`/a alias_or_expression: attack` for an alias with the name `attack`",
        "inline": false
      },
      {
        "name": "Example",
        "value": "`/a alias_or_expression: 1d6`",
        "inline": false
      },
      {
        "name": "Full documentation",
        "value": "https://github.com/twonirwana/DiscordDiceBot",
        "inline": false
      },
      {
        "name": "Discord Server for News, Help and Feature Requests",
        "value": "https://discord.gg/e43BsqKpFr",
        "inline": false
      }
    ],
    "componentRowDefinitions": [ ],
    "type": "EMBED",
    "userReference": false
  }
]


de.janno.discord.bot.command.directRoll.AliasRollCommandTest.testHelp[fr]=[
  {
    "descriptionOrContent": "Tapez `/a` et un alias ou une expression de dé. Les alias disponibles seront présentés en autocomplétion, mais il est possible de lancer toutes les expressions. `/channel_config alias` peut être utilisé pour ajouter et supprimer des alias. Le type de sortie, les images des dés, etc. peuvent être configurés avec `/channel_config save_direct_roll_config`\n```\n| Name                 | Notation   | Example            |\r\n|----------------------|------------|--------------------|\r\n| Regular Dice         | d          | d6, 3d6            |\r\n| Custom Dice          | d[x,y,...] | 3d[Head,Tail]      |\r\n| Exploding Dice       | d!         | 4d!6               |\r\n| Exploding Add Dice   | d!!        | 4d!!6              |\r\n| Keep Highest         | k          | 3d6k2              |\r\n| Keep Lowest          | l          | 3d6l2              |\r\n| Sum                  | =          | 3d6=               |\r\n| Add to List          | +          | 3d6+2d12           |\r\n| Concatenate          | _          | 3d6 _ 'dmg'        |\r\n| Negative add to List | -          | 3d6-2d12 or -2d6   |\r\n| Decimal Divide       | //         | 2/6                |\r\n| Divide               | /          | 12/6               |\r\n| Multiply             | *          | 12*d6              |\r\n| Modulo               | mod        | d6 mod 2           |\r\n| Equal Filter         | ==         | 3d6==3             |\r\n| Greater Then Filter  | >          | 3d6>3              |\r\n| Lesser Then Filter   | <          | 3d6<3              |\r\n| Greater Equal Filter | >=         | 3d6>=3             |\r\n| Lesser Equal Filter  | <=         | 3d6<=3             |\r\n| Lesser Equal Filter  | <=         | 3d6<=3             |\r\n| Count                | c          | 3d6<3c             |\r\n| Multiple Rolls       | ,          | 1d6,2d10           |\r\n| Repeat               | x          | 3x2d6              |\r\n| Repeat List          | r          | 3r2d6              |\r\n| Reroll               | rr         | 3d10 rr 1          |\r\n| Equal                | =?         | d6=?6              |\r\n| Greater              | >?         | d6>?5              |\r\n| Greater Equal        | >=?        | d6>=?5             |\r\n| Lower                | <?         | d6<?4              |\r\n| Lower Equal          | <=?        | d6<=?4             |\r\n| In                   | in         | d6 in [1/3/5]      |\r\n| And                  | &&         | d6>?1 && d6<5      |\r\n| Or                   | ||         | d6>?1 || d6<5      |\r\n| Negate               | !          | !d6>?1             |\r\n| Color                | col        | d6 col 'red'       |\r\n| Tag                  | tag        | d6 tag 'marked'    |\r\n| Brackets             | ()         | (2d4=)d6           |\r\n| Text                 | ''         | 'Damage ' + d6     |\r\n| Sort asc             | asc()      | asc(10d10)         |\r\n| Sort desc            | desc()     | desc(10d10)        |\r\n| Min                  | min()      | min(3d4, 8)        |\r\n| Max                  | max()      | max(3d4, 8)        |\r\n| Chancel              | chancel()  | chancel(8d10,10,1) |\r\n| Replace              | replace()  | replace(6d10,1,2)  |\r\n| Color On             | colorOn()  | colorOn(d6,6,'red')|\r\n| Explode              | exp()      | exp(d6,1,2)        |\r\n| If                   | if()       | if(d6=?6,'Y','N')  |\r\n| Group Count          | groupC()   | groupC(20d6)       |\r\n| Concatenate          | concat()   | concat('Att:',d20) |\r\n| Value                | val()      | val('$1',10d10)    |\r\n\n```",
    "fields": [
      {
        "name": "Exemple",
        "value": "`/a alias_or_expression: attack` for an alias with the name `attack`",
        "inline": false
      },
      {
        "name": "Exemple",
        "value": "`/a alias_or_expression: 1d6`",
        "inline": false
      },
      {
        "name": "Documentation complète",
        "value": "https://github.com/twonirwana/DiscordDiceBot",
        "inline": false
      },
      {
        "name": "Serveur Discord pour l'aide et les nouvelles",
        "value": "https://discord.gg/e43BsqKpFr",
        "inline": false
      }
    ],
    "componentRowDefinitions": [ ],
    "type": "EMBED",
    "userReference": false
  }
]


de.janno.discord.bot.command.directRoll.AliasRollCommandTest.testHelp[pt_BR]=[
  {
    "descriptionOrContent": "Digite `/a` e um alias ou expressão de dados. O alias disponível será apresentado como preenchimento automático, mas é possível rolar todas as expressões. O `/channel_config alias` pode ser usado para adicionar e excluir alias. O tipo de saída, as imagens de dados etc. podem ser configurados com `/channel_config save_direct_roll_config`\n```\n| Name                 | Notation   | Example            |\r\n|----------------------|------------|--------------------|\r\n| Regular Dice         | d          | d6, 3d6            |\r\n| Custom Dice          | d[x,y,...] | 3d[Head,Tail]      |\r\n| Exploding Dice       | d!         | 4d!6               |\r\n| Exploding Add Dice   | d!!        | 4d!!6              |\r\n| Keep Highest         | k          | 3d6k2              |\r\n| Keep Lowest          | l          | 3d6l2              |\r\n| Sum                  | =          | 3d6=               |\r\n| Add to List          | +          | 3d6+2d12           |\r\n| Concatenate          | _          | 3d6 _ 'dmg'        |\r\n| Negative add to List | -          | 3d6-2d12 or -2d6   |\r\n| Decimal Divide       | //         | 2/6                |\r\n| Divide               | /          | 12/6               |\r\n| Multiply             | *          | 12*d6              |\r\n| Modulo               | mod        | d6 mod 2           |\r\n| Equal Filter         | ==         | 3d6==3             |\r\n| Greater Then Filter  | >          | 3d6>3              |\r\n| Lesser Then Filter   | <          | 3d6<3              |\r\n| Greater Equal Filter | >=         | 3d6>=3             |\r\n| Lesser Equal Filter  | <=         | 3d6<=3             |\r\n| Lesser Equal Filter  | <=         | 3d6<=3             |\r\n| Count                | c          | 3d6<3c             |\r\n| Multiple Rolls       | ,          | 1d6,2d10           |\r\n| Repeat               | x          | 3x2d6              |\r\n| Repeat List          | r          | 3r2d6              |\r\n| Reroll               | rr         | 3d10 rr 1          |\r\n| Equal                | =?         | d6=?6              |\r\n| Greater              | >?         | d6>?5              |\r\n| Greater Equal        | >=?        | d6>=?5             |\r\n| Lower                | <?         | d6<?4              |\r\n| Lower Equal          | <=?        | d6<=?4             |\r\n| In                   | in         | d6 in [1/3/5]      |\r\n| And                  | &&         | d6>?1 && d6<5      |\r\n| Or                   | ||         | d6>?1 || d6<5      |\r\n| Negate               | !          | !d6>?1             |\r\n| Color                | col        | d6 col 'red'       |\r\n| Tag                  | tag        | d6 tag 'marked'    |\r\n| Brackets             | ()         | (2d4=)d6           |\r\n| Text                 | ''         | 'Damage ' + d6     |\r\n| Sort asc             | asc()      | asc(10d10)         |\r\n| Sort desc            | desc()     | desc(10d10)        |\r\n| Min                  | min()      | min(3d4, 8)        |\r\n| Max                  | max()      | max(3d4, 8)        |\r\n| Chancel              | chancel()  | chancel(8d10,10,1) |\r\n| Replace              | replace()  | replace(6d10,1,2)  |\r\n| Color On             | colorOn()  | colorOn(d6,6,'red')|\r\n| Explode              | exp()      | exp(d6,1,2)        |\r\n| If                   | if()       | if(d6=?6,'Y','N')  |\r\n| Group Count          | groupC()   | groupC(20d6)       |\r\n| Concatenate          | concat()   | concat('Att:',d20) |\r\n| Value                | val()      | val('$1',10d10)    |\r\n\n```",
    "fields": [
      {
        "name": "Exemplo",
        "value": "`/a alias_or_expression: attack` for an alias with the name `attack`",
        "inline": false
      },
      {
        "name": "Exemplo",
        "value": "`/a alias_or_expression: 1d6`",
        "inline": false
      },
      {
        "name": "Documentação completa",
        "value": "https://github.com/twonirwana/DiscordDiceBot",
        "inline": false
      },
      {
        "name": "Servidor do Discord Server para ajuda e notícias (inglês)",
        "value": "https://discord.gg/e43BsqKpFr",
        "inline": false
      }
    ],
    "componentRowDefinitions": [ ],
    "type": "EMBED",
    "userReference": false
  }
]